File: src/wy_qcos/tests/unit_tests/transpiler/optimizer/test_adjacent_optimization.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
# ----------------------------------------------------------------------
# Copyright© 2024-2026 China Mobile (SuZhou) Software Technology Co.,Ltd.
#
# qcos is licensed under Mulan PSL v2.
# You can use this software according to the terms and conditions
# of the Mulan PSL v2.
# You may obtain a copy of Mulan PSL v2 at:
#         http://license.coscl.org.cn/MulanPSL2
# THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS,
#     WITHOUT WARRANTIES OF ANY KIND,
# EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
# MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
# See the Mulan PSL v2 for more details.
# ----------------------------------------------------------------------

import copy
import numpy as np

from wy_qcos.transpiler.cmss.optimizer.adjacent_optimization import (
    AdjacentPhaseOptPass,
)
from wy_qcos.transpiler.cmss.circuit.dag_circuit import DAGCircuit
from wy_qcos.transpiler.cmss.common.gate_operation import (
    RX,
    RY,
    RZ,
    CRX,
    CRY,
    CRZ,
    U1,
    S,
)
from wy_qcos.tests.unit_tests.transpiler.comm import validate_optimize_result


class TestAdjacentOptimization:
    def test_adjacent_optimization(self):
        opt = AdjacentPhaseOptPass()

        ir = [
            RX([0], [0.1]),
            RX([0], [0.2]),
            RY([1], [0.1]),
            RY([1], [0.2]),
            RZ([0], [0.1]),
            RZ([0], [0.2]),
        ]
        init_ir = copy.deepcopy(ir)
        dag = DAGCircuit.ir_to_dag(ir)
        opt.run(dag)
        nodes = list(dag.topological_op_nodes())
        assert len(nodes) == 3
        assert nodes[0].op.name == "rx"
        assert np.isclose(nodes[0].op.arg_value[0], 0.3)
        assert nodes[1].op.name == "rz"
        assert np.isclose(nodes[1].op.arg_value[0], 0.3)
        assert nodes[2].op.name == "ry"
        assert np.isclose(nodes[2].op.arg_value[0], 0.3)
        validate_optimize_result(init_ir, dag)

        ir = [
            CRX([0, 1], [0.1]),
            CRX([1, 0], [0.2]),
            CRY([0, 1], [0.1]),
            CRY([0, 1], [0.2]),
            CRZ([1, 0], [0.1]),
            CRZ([1, 0], [0.2]),
        ]
        init_ir = copy.deepcopy(ir)
        dag = DAGCircuit.ir_to_dag(ir)
        opt.run(dag)
        nodes = list(dag.topological_op_nodes())
        assert len(nodes) == 4
        assert nodes[0].op.name == "crx"
        assert np.isclose(nodes[0].op.arg_value[0], 0.1)
        assert nodes[1].op.name == "crx"
        assert np.isclose(nodes[1].op.arg_value[0], 0.2)
        assert nodes[2].op.name == "cry"
        assert np.isclose(nodes[2].op.arg_value[0], 0.3)
        assert nodes[3].op.name == "crz"
        assert np.isclose(nodes[3].op.arg_value[0], 0.3)
        validate_optimize_result(init_ir, dag)

        ir = [
            RX([0], [0.1]),
            RX([0], [0.2]),
            S([1]),
            S([1]),
            U1([0], [0.1]),
            U1([0], [0.1]),
            CRY([0, 1], [0.1]),
            CRY([0, 1], [0.2]),
        ]
        init_ir = copy.deepcopy(ir)
        dag = DAGCircuit.ir_to_dag(ir)
        opt.run(dag)
        nodes = list(dag.topological_op_nodes())
        assert len(nodes) == 4
        assert nodes[0].op.name == "rx"
        assert np.isclose(nodes[0].op.arg_value[0], 0.3)
        assert nodes[1].op.name == "u1"
        assert np.isclose(nodes[1].op.arg_value[0], 0.2)
        assert nodes[2].op.name == "z"
        assert nodes[2].op.arg_value == []
        assert nodes[3].op.name == "cry"
        assert np.isclose(nodes[3].op.arg_value[0], 0.3)
        validate_optimize_result(init_ir, dag)

        # test with basis_gates
        ir = [
            RX([0], [0.1]),
            RX([0], [0.2]),
            S([1]),
            S([1]),
            U1([0], [0.1]),
            U1([0], [0.1]),
        ]
        init_ir = copy.deepcopy(ir)
        dag = DAGCircuit.ir_to_dag(ir)
        cnt = opt.run(copy.deepcopy(dag), basis_gates=["rx", "u1"])
        assert cnt == 2
        validate_optimize_result(init_ir, dag)

        cnt = opt.run(copy.deepcopy(dag), basis_gates=["rx", "u1", "rz"])
        assert cnt == 3
        validate_optimize_result(init_ir, dag)

File: src_cpp/transpiler_cpp/tests/test_adjacent_optimization.cpp
/*
 * ----------------------------------------------------------------------
 * Copyright© 2024-2026 China Mobile (SuZhou) Software Technology Co.,Ltd.
 *
 * qcos is licensed under Mulan PSL v2.
 * You can use this software according to the terms and conditions
 * of the Mulan PSL v2.
 * You may obtain a copy of Mulan PSL v2 at:
 *          http://license.coscl.org.cn/MulanPSL2
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS,
 *      WITHOUT WARRANTIES OF ANY KIND,
 * EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT,
 * MERCHANTABILITY OR FIT FOR A PARTICULAR PURPOSE.
 * See the Mulan PSL v2 for more details.
 * ----------------------------------------------------------------------
 */

#include <gtest/gtest.h>

#include <cmath>
#include <memory>
#include <string>
#include <vector>

#include "circuit/dag_circuit.h"
#include "circuit/gate_operation.h"
#include "optimizer/adjacent_optimization.h"

using namespace qcos;

TEST(AdjacentPhaseOptPassTest, MergesAdjacentParameterizedPhaseGates) {
  AdjacentPhaseOptPass optimizer;
  std::vector<std::shared_ptr<BaseOperation>> ir = {
      create_gate("rx", {0}, {M_PI / 2.0}),
      create_gate("rx", {0}, {M_PI / 4.0}),
      create_gate("ry", {0}, {M_PI / 3.0})};
  DAGCircuit dag = DAGCircuit::ir_to_dag(ir);

  EXPECT_EQ(optimizer.run(dag), 1);

  const auto counts = dag.count_ops();
  EXPECT_EQ(counts.at("rx"), 1);
  EXPECT_EQ(counts.at("ry"), 1);

  const auto nodes = dag.topological_op_nodes();
  ASSERT_EQ(nodes.size(), 2u);
  EXPECT_EQ(nodes[0]->name(), "rx");
  ASSERT_FALSE(nodes[0]->op->arg_value.empty());
  EXPECT_NEAR(nodes[0]->op->arg_value[0], 3.0 * M_PI / 4.0, 1e-9);
}

TEST(AdjacentPhaseOptPassTest, HonorsBasisFilterForParameterizedPhaseGates) {
  AdjacentPhaseOptPass optimizer;
  std::vector<std::shared_ptr<BaseOperation>> ir = {
      create_gate("rx", {0}, {M_PI / 2.0}),
      create_gate("rx", {0}, {M_PI / 4.0})};
  DAGCircuit dag = DAGCircuit::ir_to_dag(ir);

  EXPECT_EQ(optimizer.run(dag, std::set<std::string>{"ry"}), 0);

  const auto counts = dag.count_ops();
  EXPECT_EQ(counts.at("rx"), 2);
}

TEST(AdjacentPhaseOptPassTest,
     ParameterizesAndDeparameterizesDiscreteRzPhaseGates) {
  AdjacentPhaseOptPass optimizer;
  std::vector<std::shared_ptr<BaseOperation>> ir = {create_gate("s", {0}),
                                                    create_gate("s", {0})};
  DAGCircuit dag = DAGCircuit::ir_to_dag(ir);

  EXPECT_EQ(optimizer.run(dag), 1);

  const auto counts = dag.count_ops();
  EXPECT_EQ(counts.count("s"), 0u);
  EXPECT_EQ(counts.at("z"), 1);
  EXPECT_EQ(counts.count("rz"), 0u);
}